cpp/src/arrow/dataset/scanner_test.cc

```diff
@@ -319,6 +322,23 @@ class FailingFragment : public InMemoryFragment {
       return std::make_shared<InMemoryScanTask>(batches, options, self);
     });
   }
+
+  Result<RecordBatchGenerator> ScanBatchesAsync(
+      const std::shared_ptr<ScanOptions>& options) override {
+    struct {
+      Future<std::shared_ptr<RecordBatch>> operator()() {
+        if (index > 16) {
+          return Status::Invalid("Oh no, we failed!");
+        }
+        auto batch = batches[index++ % batches.size()];
+        return Future<std::shared_ptr<RecordBatch>>::MakeFinished(batch);
+      }
+      RecordBatchVector batches;
+      int index = 0;
+    } Generator;
```

Comment:
Ah, this is a neat trick for me to learn.  Is this basically creating an instance of an anonymous struct?

Reply:
Yup, it's used quite a bit in the datasets implementation (e.g. see ParquetScanTask in file_parquet.cc).